ignite/chainconfig/parser.go

```
@@ -0,0 +1,175 @@
+	ErrCouldntLocateConfig = errors.New(
+		"could not locate a config.yml in your chain. please follow the link for" +
+			"how-to: https://github.com/ignite-hq/cli/blob/develop/docs/configure/index.md")
+)
+
+// DefaultConf holds default configuration.
+var DefaultConf = v0.ConfigYaml{
+	Host: common.Host{
+		// when in Docker on MacOS, it only works with 0.0.0.0.
+		RPC:     "0.0.0.0:26657",
+		P2P:     "0.0.0.0:26656",
+		Prof:    "0.0.0.0:6060",
+		GRPC:    "0.0.0.0:9090",
+		GRPCWeb: "0.0.0.0:9091",
+		API:     "0.0.0.0:1317",
+	},
+	Build: common.Build{
+		Proto: common.Proto{
+			Path: "proto",
+			ThirdPartyPaths: []string{
+				"third_party/proto",
+				"proto_vendor",
+			},
+		},
+	},
+	Faucet: common.Faucet{
+		Host: "0.0.0.0:4500",
+	},
+}
+
+// Parse parses config.yml into UserConfig based on the version.
+func Parse(content []byte) (common.Config, error) {
+	// Read the version field
+	version, err := getConfigVersion(bytes.NewReader(content))
+	if err != nil {
+		return GetDefaultConfig(), err
+	}
+
+	conf := GetConfigInstance(version)
+	if err = yaml.NewDecoder(bytes.NewReader(content)).Decode(conf); err != nil {
+		return conf, err
+	}
+	if err = mergo.Merge(conf, conf.Default()); err != nil {
+		return GetDefaultConfig(), err
+	}
+	return conf, validate(conf)
+}
+
+// getConfigVersion returns the version in the io.Reader based on the field version.
+func getConfigVersion(r io.Reader) (string, error) {
+	var baseConf common.BaseConfigYaml
+	if err := yaml.NewDecoder(r).Decode(&baseConf); err != nil {
+		return DefaultVersion, err
+	}
+	if baseConf.Version != "" {
+		return baseConf.Version, nil
+	}
+	return DefaultVersion, nil
+}
+
+// GetConfigInstance retrieves correct config instance based on the version.
+func GetConfigInstance(version string) common.Config {
+	var config common.Config
+	var ok bool
+	if config, ok = Migration[version]; !ok {
+		// If there is no matching instance, return the config with the v0 version.
```

if there is a version that's not supported, we should return an error to the user instead of using version 0

---

still thinking...